src/pointer.rs

```diff
@@ -0,0 +1,176 @@
+        // TODO: different ordering.
+        self.data
+            .compare_exchange(current_ptr.as_usize(), new, ord, ord)
+            .map(|_| SharedPtr::from_usize(current_ptr.as_usize()))
+            .map_err(|current| (SharedPtr::from_usize(current), SharedPtr::from_usize(new)))
+    }
+}
+
+/// `SharedPtr` is a pointer which can be shared by multi-threads.
+/// `SharedPtr` can only be used with 64bit-wide pointer, and the
+/// pointer address must be 4-byte aligned.
+pub struct SharedPtr<'g, T: 'g> {
+    /// `data` is the value of the pointers.
+    /// It will be spilt into three parts:
+    /// [higher_u8, raw_pointer, lower_u2]
+    ///    8 bits     54 bits     2 bits
+    data: usize,
+    /// used for type inference.
+    _marker: PhantomData<(&'g (), *const T)>,
+}
+
+impl<T> Clone for SharedPtr<'_, T> {
+    fn clone(&self) -> Self {
+        Self {
+            data: self.data,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T> Copy for SharedPtr<'_, T> {}
+
+impl<T> SharedPtr<'_, T> {
+    /// `from_usize` creates a `SharedPtr` from a `usize`.
+    pub const fn from_usize(data: usize) -> Self {
+        SharedPtr {
+            data,
+            _marker: PhantomData,
+        }
+    }
+
+    /// `from_box` creates a `SharedPtr` from a `Box`.
+    pub fn from_box(b: Box<T>) -> Self {
+        Self::from_raw(Box::into_raw(b))
+    }
+
+    /// `from_raw` creates a `SharedPtr` from a raw pointer.
+    pub fn from_raw(raw: *const T) -> Self {
+        unsafe { Self::from_usize(std::mem::transmute::<*const T, usize>(raw)) }
+    }
+
+    /// `null` returns a null `SharedPtr`.
+    pub const fn null() -> Self {
+        Self::from_usize(0)
+    }
+
+    /// `into_box` converts the pointer into a Box<T>.
+    pub unsafe fn into_box(self) -> Box<T> {
+        Box::from_raw(self.as_mut_raw())
+    }
+
+    /// `as_usize` converts the pointer to `usize`.
+    pub const fn as_usize(self) -> usize {
+        self.data
+    }
+
+    /// `decompose_lower_u2` decomposes the pointer into two parts:
+    /// 1. the higher 62 bits
+    /// 2. the lower 2 bits
+    fn decompose_lower_u2(data: usize) -> (usize, u8) {
+        let mask: usize = 3;
+        // The unwrap is safe here, because we have mask the lower 2 bits.
+        (data & !mask, (data & mask).cast::<u8>())
+    }
+
+    /// `decompose_higher_u8` decomposes the pointer into two parts:
+    /// 1. the higher 8 bits
+    /// 2. the lower 56 bits
+    fn decompose_higher_u8(data: usize) -> (u8, usize) {
+        let mask: usize = (1_usize.overflowing_shl(56).0).overflow_sub(1);
+        // The conversion is safe here, because we have shifted 56 bits.
+        (data.overflow_shr(56).cast::<u8>(), data & mask)
+    }
+
+    /// `decompose` decomposes the pointer into three parts:
+    /// 1. the higher 8 bits
+    /// 2. the raw pointer
+    /// 3. the lower 2 bits
+    #[allow(clippy::as_conversions)]
+    pub fn decompose(self) -> (u8, *const T, u8) {
```

Comment:
should this function be marked as unsafe?

Reply:
IMHO, generate a raw pointer is safe. The unsafe operation is to dereferencing a raw pointer or convert a raw pointer to a reference, e.g.
```rust
let raw_ptr = 0_usize as *const T; // this is safe
let _ = unsafe {*raw_ptr}; // this is unsafe
let _ = unsafe {raw_ptr.as_ref()}; // this is unsafe
```